File: ios-native/Sprout/Design/Tilt.swift
import CoreMotion
import Observation
import SwiftUI
import UIKit

/// Наклон телефона — источник параллакса узора, и он же слышит тряску.
///
/// Один на приложение: датчик у телефона один; пока ни один экран не смотрит,
/// он выключен. Отсчёт — от медленно ползущей «привычной» точки, а не от
/// вертикали: так узор не упирается в край, как бы ни держали телефон, и
/// калибровка не нужна.
@Observable
final class Tilt {
    static let shared = Tilt()

    private(set) var shift: CGSize = .zero

    /// Расхождение каждого слоя фигурок с узором, огрублённое до полупункта,
    /// — см. `Sway`. Огрубление — чтобы холст не пересобирался на каждую
    /// сотую пункта; присвоение под проверкой, потому что `@Observable` будит
    /// зависимых на каждую запись.
    private(set) var lag: [CGSize] = []

    /// Меняется в покое — и с ним жребий, каким фигуркам какая вязкость.
    private(set) var era = 0

    private static let swayStep: CGFloat = 0.5

    @ObservationIgnored private var places: [CGSize] = []

    @ObservationIgnored private var calm = 0.0

    /// Настройку кладёт фон: датчик не знает про хранилище. Выключенный
    /// параллакс не гасит датчик — тряску выключать не просили; узор встаёт
    /// на место сразу.
    var parallax = true {
        didSet {
            guard parallax != oldValue, !parallax else { return }
            shift = .zero
            lag = []
            places = []
            calm = 0
        }
    }

    /// Полный размах — примерно за 17°.
    private static let gain = Metrics.parallax / 0.3

    /// Привычная точка возвращается в середину примерно за полторы секунды.
    private static let baseEase = 0.011

    /// Сглаживает дрожь датчика.
    private static let ease = 0.12

    @ObservationIgnored private let motion = CMMotionManager()
    @ObservationIgnored private var watchers = 0
    @ObservationIgnored private var base: (x: Double, z: Double)?

    @ObservationIgnored private var shaken = 0.0
    @ObservationIgnored private var lastSample: Date?

    /// Порог тряски в долях тяжести (`userAcceleration`): при ходьбе там до
    /// трети, так что случайно не набрать.
    private static let shakeForce = 1.5

    /// Сколько надо натрясти в сумме, а не подряд: за взмах датчик выходит за
    /// порог дважды. Спускается вдвое быстрее, чем копится.
    private static let shakeSeconds = 2.2
    private static let shakeFade = 2.0

    private init() {}

    func watch() {
        watchers += 1
        guard watchers == 1,
              // При «Уменьшении движения» датчик не включается вовсе.
              !UIAccessibility.isReduceMotionEnabled,
              motion.isDeviceMotionAvailable
        else { return }

        base = nil
        motion.deviceMotionUpdateInterval = 1.0 / 60
        shaken = 0
        lastSample = nil
        motion.startDeviceMotionUpdates(to: .main) { [weak self] data, _ in
            guard let self, let data else { return }
            self.step(x: data.gravity.x, z: data.gravity.z)
            self.feel(data.userAcceleration)
        }
    }

    func unwatch() {
        watchers = max(0, watchers - 1)
        guard watchers == 0 else { return }
        motion.stopDeviceMotionUpdates()
        base = nil
        shaken = 0
        lastSample = nil
        shift = .zero
        lag = []
        places = []
        calm = 0
    }

    /// Сила тяжести в осях телефона: `x` — крен, `z` — наклон к себе и от
    /// себя; третья ось меняется вместе с `z`.
    private func step(x: Double, z: Double) {
        guard let seen = base else {
            base = (x, z)
            return
        }
        base = (seen.x + (x - seen.x) * Self.baseEase,
                seen.z + (z - seen.z) * Self.baseEase)

        guard parallax else { return }

        // Против наклона — так узор читается лежащим за экраном.
        let target = CGSize(width: limit(-(x - seen.x) * Self.gain),
                            height: limit(-(z - seen.z) * Self.gain))
        shift = CGSize(
            width: shift.width + (target.width - shift.width) * Self.ease,
            height: shift.height + (target.height - shift.height) * Self.ease)

        if places.count != Sway.eases.count { places = Sway.rest(at: shift) }
        places = Sway.settle(places, toward: target)
        let fresh = Sway.lag(places, behind: shift).map(Self.rough)
        if fresh != lag { lag = fresh }

        if fresh.allSatisfy({ $0.width == 0 && $0.height == 0 }) {
            calm += motion.deviceMotionUpdateInterval
            if calm >= Sway.shuffleSeconds {
                calm = 0
                era &+= 1
            }
        } else {
            calm = 0
        }
    }

    private static func rough(_ size: CGSize) -> CGSize {
        CGSize(width: (size.width / swayStep).rounded() * swayStep,
               height: (size.height / swayStep).rounded() * swayStep)
    }

    /// Трясут ли телефон — тем же `CMDeviceMotion`, что идёт ради параллакса.
    /// См. `Frolic`.
    private func feel(_ push: CMAcceleration) {
        let now = Date()
        // Промежуток по часам, а не по числу отсчётов: они задерживаются.
        // Сверху ограничен — приложение могло стоять в фоне.
        let gap = min(max(now.timeIntervalSince(lastSample ?? now), 0), 0.1)
        lastSample = now

        let force = sqrt(push.x * push.x + push.y * push.y + push.z * push.z)
        guard force > Self.shakeForce else {
            shaken = max(0, shaken - gap * Self.shakeFade)
            return
        }
        shaken += gap
        guard shaken >= Self.shakeSeconds else { return }
        shaken = 0
        // Отклик — только с настоящим началом, иначе бил бы каждые две
        // секунды, пока трясут.
        guard Frenzy.shared.begin() else { return }
        Task { @MainActor in Feel.frenzy() }
    }

    private func limit(_ value: Double) -> CGFloat {
        CGFloat(min(max(value, -Metrics.parallax), Metrics.parallax))
    }
}

File: android/app/src/main/kotlin/com/ledro6/sprout/platform/Tilt.kt
package com.ledro6.sprout.platform

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.SystemClock
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.ledro6.sprout.design.Effects
import com.ledro6.sprout.design.Metrics
import com.ledro6.sprout.model.Sway
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sqrt

/**
 * Наклон телефона — источник параллакса узора, и он же слышит тряску.
 * Отсчёт — от медленно ползущей «привычной» точки, а не от вертикали: так
 * узор не упирается в край, как бы ни держали телефон. Датчик включён, только
 * пока фон на экране. Оси и знаки — как у iPhone: сила тяжести в долях `g`,
 * от экрана к глазу — минус.
 */
object Tilt : SensorEventListener {
    /** Общий сдвиг узора, dp. */
    var shiftX by mutableStateOf(0f)
        private set
    var shiftY by mutableStateOf(0f)
        private set

    /** Расхождение слоёв фигурок, огрублённое до полупункта, — пары x, y. */
    var lag: List<Float> by mutableStateOf(emptyList())
        private set

    /** Меняется в покое — и с ним жребий слоёв. */
    var era by mutableIntStateOf(0)
        private set

    var parallax = true
        set(value) {
            if (field == value) return
            field = value
            if (!value) rest()
        }

    private var manager: SensorManager? = null
    private var watchers = 0
    private var baseX = 0.0
    private var baseZ = 0.0
    private var based = false
    private val places = DoubleArray(Sway.eases.size * 2)
    private var placed = false
    private var calm = 0.0
    private var shaken = 0.0
    private var lastSample = 0L
    private var lastShake = 0L

    private const val GAIN = Metrics.PARALLAX / 0.3
    private const val BASE_EASE = 0.011
    private const val EASE = 0.12
    private const val SHAKE_FORCE = 1.5
    private const val SHAKE_SECONDS = 2.2
    private const val SHAKE_FADE = 2.0

    fun watch(context: Context) {
        watchers += 1
        if (watchers != 1) return
        val sensors = context.getSystemService(Context.SENSOR_SERVICE) as? SensorManager ?: return
        manager = sensors
        based = false
        shaken = 0.0
        lastSample = 0L
        val gravity = sensors.getDefaultSensor(Sensor.TYPE_GRAVITY)
            ?: sensors.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
        gravity?.let { sensors.registerListener(this, it, SensorManager.SENSOR_DELAY_GAME) }
        sensors.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION)?.let {
            sensors.registerListener(this, it, SensorManager.SENSOR_DELAY_GAME)
        }
    }

    fun unwatch() {
        watchers = max(0, watchers - 1)
        if (watchers != 0) return
        runCatching { manager?.unregisterListener(this) }
        manager = null
        rest()
    }

    private fun rest() {
        based = false
        placed = false
        shiftX = 0f
        shiftY = 0f
        lag = emptyList()
        calm = 0.0
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) = Unit

    override fun onSensorChanged(event: SensorEvent) {
        when (event.sensor.type) {
            Sensor.TYPE_GRAVITY, Sensor.TYPE_ACCELEROMETER ->
                step(-event.values[0] / SensorManager.GRAVITY_EARTH.toDouble(), -event.values[2] / SensorManager.GRAVITY_EARTH.toDouble())
            Sensor.TYPE_LINEAR_ACCELERATION -> feel(event.values[0], event.values[1], event.values[2])
        }
    }

    private fun step(x: Double, z: Double) {
        if (!based) {
            baseX = x
            baseZ = z
            based = true
            return
        }
        val seenX = baseX
        val seenZ = baseZ
        baseX += (x - seenX) * BASE_EASE
        baseZ += (z - seenZ) * BASE_EASE
        if (!parallax) return
        val targetX = limit(-(x - seenX) * GAIN)
        val targetY = limit(-(z - seenZ) * GAIN)
        val nextX = shiftX + (targetX - shiftX) * EASE
        val nextY = shiftY + (targetY - shiftY) * EASE
        shiftX = nextX.toFloat()
        shiftY = nextY.toFloat()
        if (!placed) {
            for (i in Sway.eases.indices) {
                places[2 * i] = nextX
                places[2 * i + 1] = nextY
            }
            placed = true
        }
        Sway.settle(places, targetX, targetY)
        val fresh = List(places.size) { rough(Sway.hold(places[it] - if (it % 2 == 0) nextX else nextY)) }
        if (fresh != lag) lag = fresh
        if (fresh.all { it == 0f }) {
            calm += 1.0 / 50
            if (calm >= Sway.SHUFFLE_SECONDS) {
                calm = 0.0
                era += 1
            }
        } else {
            calm = 0.0
        }
    }

    private fun rough(value: Double): Float = ((value / 0.5).roundToInt() * 0.5).toFloat()

    private fun limit(value: Double): Double = min(max(value, -Metrics.PARALLAX), Metrics.PARALLAX)

    /** Трясут ли телефон: сила сверх 1.5 g, накопленная за 2.2 секунды. */
    private fun feel(x: Float, y: Float, z: Float) {
        val now = SystemClock.elapsedRealtime()
        val gap = if (lastSample == 0L) 0.0 else min(max((now - lastSample) / 1000.0, 0.0), 0.1)
        lastSample = now
        val force = sqrt((x * x + y * y + z * z).toDouble()) / SensorManager.GRAVITY_EARTH
        if (force <= SHAKE_FORCE) {
            shaken = max(0.0, shaken - gap * SHAKE_FADE)
            return
        }
        shaken += gap
        if (shaken < SHAKE_SECONDS) return
        shaken = 0.0
        if (now - lastShake < 1000) return
        lastShake = now
        if (Effects.frenzy()) Feel.frenzy()
    }
}
